test/remote/redis_tests.js

```diff
@@ -0,0 +1,226 @@
+    return P.all([ redis.get('foo'), redis.get('foo') ])
+      .then(results => results.forEach(result => assert.equal(result, 'bar')))
+  })
+})
+
+describe('concurrent sets:', () => {
+  before(() => {
+    return P.all([ redis.set('foo', '1'), redis.set('foo', '2') ])
+  })
+
+  it('data was set', () => {
+    return redis.get('foo')
+      .then(result => assert.ok(result === '1' || result === '2'))
+  })
+})
+
+describe('redis.del:', () => {
+  before(() => {
+    return redis.del('foo')
+  })
+
+  it('data was deleted', () => {
+    return redis.get('foo')
+      .then(result => assert.ok(result === null))
+  })
+})
+
+describe('redis.update:', () => {
+  before(() => {
+    return redis.set('foo', 'bar')
+      .then(() => redis.update('foo', oldValue => `${oldValue}2`))
+  })
+
+  it('data was set', () => {
+    return redis.get('foo')
+      .then(result => assert.equal(result, 'bar2'))
+  })
+})
+
+describe('update non-existent key:', () => {
+  before(() => {
+    return redis.del('wibble')
+      .then(() => redis.update('wibble', () => 'blee'))
+  })
+
+  it('data was set', () => {
+    return redis.get('wibble')
+      .then(result => assert.equal(result, 'blee'))
+  })
+})
+
+describe('update existing key to falsey value:', () => {
+  before(() => {
+    return redis.update('wibble', () => '')
+  })
+
+  it('data was deleted', () => {
+    return redis.get('wibble')
+      .then(result => assert.ok(result === null))
+  })
+})
+
+describe('concurrent updates of the same key:', () => {
+  const errors = []
+  let winner
+
+  before(() => {
+    let resolve, sum = 0
+    const synchronisationPromise = new P(r => resolve = r)
+
+    return P.all([ 1, 2 ].map(value => {
+      return redis.update('foo', createUpdateHandler(value))
+        .then(() => winner = value)
+        .catch(error => errors.push(error))
+    }))
+
+    function createUpdateHandler (value) {
+      return () => {
+        sum += value
+        if (sum === 3) {
+          resolve()
+        }
+        return synchronisationPromise
+          .then(() => value)
+      }
+    }
+  })
+
+  it('one update failed', () => {
+    assert.equal(errors.length, 1)
+    assert.equal(errors[0].message, 'Unspecified error')
+    assert.equal(errors[0].errno, 999)
+  })
+
+  it('the other update completed successfully', () => {
+    return redis.get('foo')
+      .then(result => assert.equal(result, winner))
+  })
+})
+
+describe('concurrent updates of different keys:', () => {
+  before(() => {
+    let resolve, values = ''
+    const synchronisationPromise = new P(r => resolve = r)
+
+    return P.all([
+      redis.update('foo', createUpdateHandler('bar')),
+      redis.update('baz', createUpdateHandler('qux'))
+    ])
+
+    function createUpdateHandler (value) {
+      return () => {
+        values += value
+        if (values.length === 6) {
+          resolve()
+        }
+        return synchronisationPromise
+          .then(() => value)
+      }
+    }
+  })
```

Comment:
👍

Reply:
:smile: Yep, as usual your comment earlier about these tests was impressively prescient. The introduction of disposers meant the initial watches were no longer occurring in the current tick, so I had to follow your advice and enforce concurrency from the handler.